### ISSUE TITLE
incorrect SNI(Server Name Indication) of TLS.

### ISSUE THREAD

**Issue by goat3** [pocoproject/poco#2056]
#### Expected behavior
I understand that the domain name should be set into server_name(Server Name Indication) of TLS Extensions in the Client Hello message.

#### Actual behavior
At first time, server_name is not set.
After second times, foreign global IP address is set into server_name instead of domain name.
(I saw it at the packet by WireShark.)

In the result, HTTPS client made by Poco get the TLS response(Server Hello Message) with "Unrecognized Name" alert from some HTTPS server. And client fails to communicate with HTTPS server.

#### Steps to reproduce the problem
(1) Create Poco::Net::HTTPSClientSession with Poco::Net::Context.
(2) Call Poco::Net::HTTPSClientSession::setHost() with domain name.
(3) Call Poco::Net::HTTPSClientSession::setPort().
(4) Call Poco::Net::HTTPSClientSession::sendRequest() with Poco::Net::HTTPRequest.
    --> The server_name is not set in this first time.
(5) Call Poco::Net::HTTPSClientSession::receiveResponse().
(6) Execute No(4) and No(5) again.
   --> The foreign global IP address is set into server_name in this second time.

#### POCO version
poco-1.6.1-all
(latest poso-1.8.0-all may has same problem, because we can see same codes in this version)

#### Compiler and version
g++ 4.9.3

#### Operating system and version
Linux Ubuntu 14.04 (Kernel 3.14)

#### Other relevant information
The following code is in poco-1.6.1-all/NetSSL_OpenSSL/src/SecureSocketImpl.cpp.
SSL_set_tlsext_host_name() is called with _peerHostName.
In the result of first communication, _peerHostName will get global IP address by using ::getpeername().
Therefore global IP address is set into server_name after second times.

~~~~
void SecureSocketImpl::connectSSL(bool performHandshake)
{
	poco_assert (!_pSSL);
	poco_assert (_pSocket->initialized());
	
	BIO* pBIO = BIO_new(BIO_s_socket());
	if (!pBIO) throw SSLException("Cannot create SSL BIO object");
	BIO_set_fd(pBIO, static_cast<int>(_pSocket->sockfd()), BIO_NOCLOSE);

	_pSSL = SSL_new(_pContext->sslContext());
	if (!_pSSL) 
	{
		BIO_free(pBIO);
		throw SSLException("Cannot create SSL object");
	}
	SSL_set_bio(_pSSL, pBIO, pBIO);
	
#if OPENSSL_VERSION_NUMBER >= 0x0908060L && !defined(OPENSSL_NO_TLSEXT)
	if (!_peerHostName.empty())
	{
		SSL_set_tlsext_host_name(_pSSL, _peerHostName.c_str());
	}
#endif

	if (_pSession)
	{
		SSL_set_session(_pSSL, _pSession->sslSession());
	}
	
	try
	{
		if (performHandshake && _pSocket->getBlocking())
		{
			int ret = SSL_connect(_pSSL);
			handleError(ret);
			verifyPeerCertificate();
		}
		else
		{
			SSL_set_connect_state(_pSSL);
			_needHandshake = true;
		}
	}
	catch (...)
	{
		SSL_free(_pSSL);
		_pSSL = 0;
		throw;
	}
}
~~~~

**Comment by aleks-f** [comment]
Yeah, the `SSL_set_tlsext_host_name` call seems to be of no use. But getting involved in name resolution here means potentially blocking for a long time. @obiltschnig

**Comment by obiltschnig** [comment]
I cannot reproduce this. For me, SSL_set_tlsext_host_name() is called with the correct hostname specified to HTTPSClientSession. Please show exactly how you're setting up the HTTPSClientSession.

**Comment by obiltschnig** [comment]
@alex The call to `SSL_set_tlsext_host_name()` is correct. Problem is that the host name is apparently not being passed to `SecureSocketImpl`.

**Comment by goat3** [comment]
Thank you for quick reply.
I got the problem on the specific micro server(execution environment), but I could not see the problem on VM(Debian, development environment) with same poco-1.6.1-all.
I have to confirm again, and prepare detail procedure.
Please give me few time.

**Comment by goat3** [comment]
I confirmed again on VM(Hyper-V) and I met same problem on this general environment, too.
I report additional information that this problem only occurrs when proxy is not used.
I attached reproducible code.

Please attach debugger and see _peerHostName value at the SecureSocketImpl::connectSSL().
You can see null string in first time, and you can see global IP address in second time.
And you can see these values at server_name of TLS Extensions in the Client Hello message, too.

#### POCO version
poco-1.6.1-all

#### Compiler and version
g++ 4.7

#### Operating system and version
Linux Ubuntu 7.8 (Kernel 3.2.0-4-686-pae) on i686

#### Steps to reproduce the problem
~~~~
#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <sstream>
#include <auto_ptr.h>
#include <Poco/Net/HTTPRequest.h>
#include <Poco/Net/HTTPResponse.h>
#include <Poco/Net/HTTPClientSession.h>
#include <Poco/Net/HTTPSClientSession.h>
#include <Poco/Net/Context.h>
#include <Poco/StreamCopier.h>

using namespace std;

// Problem occurrs when proxy is not used
#define _USE_PROXY 0

int main()
{
	Poco::Net::Context::Ptr pSSLContext( new Poco::Net::Context(
		Poco::Net::Context::CLIENT_USE,
		"", "", "", Poco::Net::Context::VERIFY_NONE, 9, false,
		"ALL:!ADH:!LOW:!EXP:!MD5:@STRENGTH"
	));

	auto_ptr<Poco::Net::HTTPClientSession> pSession(
		 new Poco::Net::HTTPSClientSession(pSSLContext) );
	pSession->setHost( "google.com" );
	pSession->setPort( 443 );
	pSession->setTimeout( Poco::Timespan(5, 0) );

#if _USE_PROXY
	// No problem in using proxy
	pSession->setProxy( "www-proxy.***.com", 80 );
#endif

	for ( int i=0; i<2; ++i )
	{
		string path = "https://google.com";
		string body;

		Poco::Net::HTTPRequest req(
			Poco::Net::HTTPRequest::HTTP_GET,
			path,
			Poco::Net::HTTPMessage::HTTP_1_1
		);
		req.setKeepAlive( true );
		req.set( "Accept-Charset", "UTF-8" );

		req.setContentLength( body.size() );
		std::ostream& ss = pSession->sendRequest( req );
		ss << body;

		Poco::Net::HTTPResponse res;
		std::istream& rs = pSession->receiveResponse( res );

		stringstream ostr;
		Poco::StreamCopier::copyStream( rs, ostr );
		cout <<  "response: " << ostr.str() << endl;
	}

	return 0;
}
~~~~

**Comment by goat3** [comment]
I attached screenshots.

#### _peerHostName value in first time
> ![peerhostname_first_time](https://user-images.githubusercontent.com/34646013/34192007-712edfe6-e58e-11e7-8f52-f5cd6c5f3601.jpg)

#### _peerHostName value in second time
> ![peerhostname_second_time](https://user-images.githubusercontent.com/34646013/34192011-76f4a5f0-e58e-11e7-8a26-57fd7937dc24.jpg)

#### server_name at Client Hello message by WireShark
> ![tls client hello message pakcet](https://user-images.githubusercontent.com/34646013/34191735-c264d124-e58c-11e7-8fe8-61de2ce1ba4c.jpg)

**Comment by obiltschnig** [comment]
At least with 1.8.1 I cannot reproduce this with your sample. The server name is correctly set in the first request. After checking the source history, it seems that this issue was fixed in release 1.7.4.

**Comment by goat3** [comment]
I confirmed that this problem was fixed in poco-1.8.0.1-all.
Thank you for your help !!
